examples/pytorch/aml.py

```diff
@@ -0,0 +1,156 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""
+This example demonstrates how to use ray to submit a job to Azure ML using ray-on-aml.
+
+Multi-node users need to follow these steps,
+0. https://learn.microsoft.com/en-us/azure/machine-learning/how-to-secure-training-vnet?view=azureml-api-2&tabs=cli%2Crequired
+1. Create a virtual network and add your compute to it.
+2. Create a network security group.
+3. Manually include pip packages shown in ray_on_aml + others required
+4. For multi-node, this script needs to be run in azure ml terminal not local.
+"""
+from time import sleep
+
+import ray
+from ray.train.torch import TorchTrainer
+from ray.air import session
+from ray.air.config import ScalingConfig, RunConfig
+
+from deepgnn.graph_engine.data.ppi import PPI
+from deepgnn.graph_engine.snark.distributed import Server, Client as DistributedClient
+
+
+@ray.remote
+class ServerLock:
+    """Lock for server."""
+
+    def __init__(self, size):
+        """Initialize server lock."""
+        self.cache = {}
+        self.locks = [False for _ in range(size)]
+
+    def set_lock(self, i):
+        """Set lock i."""
+        self.locks[i] = True
+
+    def release_lock(self, i):
+        """Release lock i."""
+        self.locks[i] = False
+
+    def wait_set(self):
+        """See if all locks set."""
+        return all(self.locks)
+
+    def wait_released(self):
+        """See if all locks released."""
+        return all([not lock for lock in self.locks])
+
+    def put(self, x, y):
+        """Put value y at x."""
+        self.cache[x] = y
+
+    def get(self, x):
+        """Get value x."""
+        return self.cache.get(x)
+
+
+def set_lock(server_lock, rank, address):
+    """Set lock at rank, then halt until all servers locked."""
+    server_lock.put.remote(rank, address)
+    server_lock.set_lock.remote(rank)
+    while not ray.get(server_lock.wait_set.remote()):
+        sleep(1)
+
+
+def release_lock(server_lock, rank):
+    """Release lock at rank, then halt until all servers released."""
+    server_lock.release_lock.remote(session.get_world_rank())
+    while not ray.get(server_lock.wait_released.remote()):
+        sleep(1)
+
+
+def train_func(config: dict):
+    """Training loop for ray trainer."""
+    ppi = PPI(num_partitions=session.get_world_size())
+
+    address = (
+        f"{ray._private.services.get_node_ip_address()}:999{session.get_world_rank()}"
+    )
+
+    set_lock(server_lock, session.get_world_rank(), address)
+
+    address = ray.get(
+        [server_lock.get.remote(i) for i in range(session.get_world_size())]
+    )
+
+    Server(
```

Comment:
1. Why do we need locks if the server is created before client in the training loop?
2. Where do you shutdown server?
3. Could you separate "business" logic of using a graph for querying with infrastructure logic of setting up multiple servers and connecting a client to them when they are ready.

Reply:
1. Locks are since a server is created in each worker/node so they will not necessarily be created before every client.
2. We wait for all locks to be released at the end, then they go out of scope and get closed.
3. Could you clarify // show code for this one.

Reply:
1. Locks should have a wait before a client is created, because other workers might start late and a client will fail to connect to the server located on it.
2. Thanks!
3. You probably want to have something like this: https://docs.ray.io/en/latest/ray-core/patterns/tree-of-actors.html A supervisor starts GE servers in a constructor and then calls fit for every worker it is responsible for in a separate method.

Reply:
1. Updated.
3. What is the benefit of this added complexity? My simple solution seems to work.